File: LeetCode/Graph/DFS/JS/Leetcode784-letter-case-permutation.js
const isdigit = (ch) => {
  return parseFloat(ch).toString() === "NaN" ? false : true;
}
var letterCasePermutation = function(s) {
  let [n, ans] = [s.length, []];
  s = [...s];
  let dfs = (i) => {
      if(i == n) {
          ans.push(s.join(''));
          return;
      }
      dfs(i + 1);
      if(!isdigit(s[i])) {
          s[i] = String.fromCharCode(s[i].charCodeAt() ^ 32);
          dfs(i + 1);
          s[i] = String.fromCharCode(s[i].charCodeAt() ^ 32);
      }
  }
  dfs(0);
  return ans;
};

File: LeetCode/Graph/DFS/Python/Leetcode784-letter-case-permutation.py
class Solution:
    def letterCasePermutation(self, s: str) -> List[str]:
        n, ans = len(s), []
        s = list(s)
        def dfs(i):
            nonlocal n
            if i == n: 
                ans.append(''.join(s))
                return 
            dfs(i + 1)
            if not s[i].isdigit():
                s[i] = s[i].swapcase()
                dfs(i + 1)
                s[i] = s[i].swapcase()
        dfs(0)
        return ans
